packages/docusaurus-plugin-client-redirects/src/index.ts

```diff
@@ -0,0 +1,144 @@
+  redirectPageContent: string;
+  redirectAbsoluteFilePath: string;
+};
+
+type PluginContext = {
+  props: Props;
+  options: PluginOptions;
+  redirectsCreators: RedirectsCreator[];
+};
+
+export default function pluginClientRedirectsPages(
+  _context: LoadContext,
+  opts: Partial<PluginOptions>,
+): Plugin<unknown> {
+  const options = {...DEFAULT_OPTIONS, ...opts};
+  return {
+    name: 'docusaurus-plugin-client-redirects',
+    async postBuild(props: Props) {
+      const redirectsCreators: RedirectsCreator[] = buildRedirectCreators(
+        options,
+      );
+
+      const pluginContext: PluginContext = {props, options, redirectsCreators};
+      // Process in 2 steps, to make code more easy to test
+      const redirects: RedirectMetadata[] = collectRoutePathRedirects(
+        pluginContext,
+      );
+
+      console.log('redirects=', redirects);
+
+      await writeRedirectFiles(redirects);
+    },
+  };
+}
+
+function buildRedirectCreators(options: PluginOptions): RedirectsCreator[] {
+  const noopRedirectCreator: RedirectsCreator = (_routePath: string) => [];
+  return [
+    fromExtensionsRedirectCreator(options.fromExtensions),
+    toExtensionsRedirectCreator(options.toExtensions),
+    options.createRedirects ?? noopRedirectCreator,
+  ];
+}
+
+function collectRoutePathRedirects(
+  pluginContext: PluginContext,
+): RedirectMetadata[] {
+  return flatten(
+    pluginContext.redirectsCreators.map((redirectCreator) => {
+      return createRoutesPathsRedirects(redirectCreator, pluginContext);
+    }),
+  );
+}
+
+// Create all redirects for a list of route path
+function createRoutesPathsRedirects(
+  redirectCreator: RedirectsCreator,
+  pluginContext: PluginContext,
+): RedirectMetadata[] {
+  return flatten(
+    pluginContext.props.routesPaths.map((routePath) =>
+      createRoutePathRedirects(routePath, redirectCreator, pluginContext),
+    ),
+  );
+}
+
+// Create all redirects for a single route path
+function createRoutePathRedirects(
+  routePath: string,
+  redirectCreator: RedirectsCreator,
+  {props}: PluginContext,
+): RedirectMetadata[] {
+  const {siteConfig, outDir} = props;
+
+  // TODO do we receive absolute urls???
+  if (!path.isAbsolute(routePath)) {
+    return [];
+  }
+
+  // TODO addTrailingSlash ?
+  const toUrl = addTrailingSlash(`${siteConfig.url}${routePath}`);
+
+  const redirectPageContent = createRedirectPageContent({toUrl});
+
+  const fromRoutePaths: string[] = redirectCreator(routePath) ?? [];
+
+  return fromRoutePaths.map((fromRoutePath) => {
+    const redirectAbsoluteFilePath = path.join(
+      outDir,
+      getFilePathForRoutePath(fromRoutePath),
+    );
+    return {
+      fromRoutePath,
+      toRoutePath: routePath,
+      toUrl,
+      redirectPageContent,
+      redirectAbsoluteFilePath,
+    };
+  });
+}
+
+async function writeRedirectFiles(redirects: RedirectMetadata[]) {
+  async function writeRedirectFile(redirect: RedirectMetadata) {
+    try {
+      await fs.writeFile(
```

Comment:
Please use sync file write (see https://github.com/facebook/docusaurus/pull/2530 for more details and examples).

Reply:
This shouldn't be a problem in this case, because I'm using fs-extra and awaiting the file writes

Reply:
Alright, if so, otherwise in certain cases the website building may be finished before the redirect files are generated.